week3/src/Fight.js

```diff
@@ -0,0 +1,193 @@
+import { useEffect } from "react";
+import styled from "styled-components";
+import sword from "./img/sword.png";
+import crown from "./img/crown.png";
+import RetryGame from "./RetryGame";
+import ShowImage from "./ShowImage.js";
+import { keyframes } from "styled-components";
+
+const PageTitle = styled.h1`
+  text-align: center;
+  margin: 0;
+  padding-top: 20px;
+  padding-bottom: 20px;
+  color: black;
+  display: flex;
+  justify-content: center;
+  font-size: 30px;
+`;
+
+const ImgContainer = styled.div`
+  display: flex;
+  justify-content: center;
+  margin: 40px;
+  margin-top: 0;
+`;
+
+const EnlangeImage = keyframes`
+  0% {
+    transform: scale(1);
+  }
+  100% {
+    transform: scale(2.0);
+  }
+  `;
+
+const ImgContents = styled.img`
+  overflow: hidden;
+  border-radius: 250px;
+  border: 12px solid rgb(255, 186, 0);
+  background-color: black;
+  margin: 0 5px;
+  height: 500px;
+  width: 500px;
+`;
+
+const RoundTitle = styled.h2`
+  text-align: center;
+  margin: 0;
+  /* padding-top: 20px; */
+  color: black;
+  display: flex;
+  justify-content: center;
+  font-size: 60px;
+`;
+
+const RoundText = styled.span`
+  font-size: 80px;
+  color: red;
+`;
+
+const OtherText = styled.span`
+  padding-top: 15px;
+`;
+
+const CrownImg = styled.img`
+  width: 100px;
+  height: 100px;
+  position: absolute;
+  left: 47%;
+  top: 10%;
+`;
+const VsContainer = styled.div`
+  position: relative;
+  display: flex;
+  align-items: center;
+`;
+
+const LeftSwordCross = keyframes`
+  0% {
+    margin-top: 0;
+  }
+  50% {
+    margin-top:-100px;
+    margin-left: 20px;
+  }
+  100% {
+    margin-top: 0;
+  }
+  `;
+
+const RightSwordCross = keyframes`
+0% {
+  margin-top: 0;
+}
+50% {
+  margin-top:-100px;
+  margin-right: 20px;
+}
+100% {
+  margin-top: 0;
+}
+`;
+
+const LeftSword = styled.img`
+  /* position: absolute; */
+  height: 100px;
+  width: 100px;
+  animation: ${LeftSwordCross} 2s 1s infinite;
+  transform: translate(-20px) rotate(5deg);
+`;
+
+const RightSword = styled.img`
+  position: absolute;
+  left: 20px;
+  height: 100px;
+  width: 100px;
+  transform: scaleX(-1) rotate(5deg);
+  animation: ${RightSwordCross} 2s 1s infinite;
+`;
+
+const FightGame = ({
+  fighter,
+  setFighter,
+  winImg,
+  setWinImg,
+  eventHandle,
+  round,
+  setRound,
+}) => {
+  const displayImg = [fighter[0], fighter[1]];
+  useEffect(() => {
+    if (fighter.length === 0) {
+      setFighter(winImg);
+      setWinImg([]);
+    }
+  });
```

Comment:
```suggestion
  useEffect(() => {
    if (fighter.length === 0) {
      setFighter(winImg);
      setWinImg([]);
    }
  },[fighter]);
```
`useEffect(함수,[]);`는 기본적으로 LifeCycle에 따라 특정 시점에 함수를 실행시킵니다! 여기서 함수 실행시킬 조건이 []에 들어가고, 실행될 함수가 첫번재 인자로 들어갑니다. 만약 지금처럼 두번째 인자를 비워두면 맨처음에 렌더 되었을 때, 재렌더링 될 때, 소멸 될 때 모든 시점에 함수가 실행되어 불필요한 때에도 useEffect 내부 로직이 실행되게 됩니다! 그래서 필요한 시점에만 useEffect내부 함수가 호출될 수 있도록 두번째 인자를 정의해주는 것이 좋을 것 같아요.

Reply:
아래에서 사용한 useEffect의 경우 `winImg`값이 변화할 때 `useEffect` 내부 함수를 실행시키는 것이니까 배열 내부 값을 `[winImg]` 값으로 해주었는데, 이 부분은 `fighter`의 값이 변화할 때 실행시켜줘야 하는 로직이니까  useEffect의 두번째 인자 배열 값을 `[fighter]`로 넣어주면 좋을 것 같습니다 : )